drivers/serial/uart_emul.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2023 Fabian Blatz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_uart_emul
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/drivers/serial/uart_emul.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_emul, CONFIG_UART_LOG_LEVEL);
+
+struct uart_emul_config {
+};
+
+struct uart_emul_data {
+	struct uart_config cfg;
+
+	struct ring_buf rx_rb;
+	uint8_t rx_buf[CONFIG_UART_EMUL_BUFFER_SIZE];
+	struct k_spinlock rx_lock;
+
+	struct ring_buf tx_rb;
+	uint8_t tx_buf[CONFIG_UART_EMUL_BUFFER_SIZE];
+	struct k_spinlock tx_lock;
+};
+
+int uart_emul_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	struct uart_emul_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	if (ring_buf_is_empty(&data->rx_rb)) {
+		LOG_DBG("Rx buffer is empty");
+		return -1;
+	}
+	key = k_spin_lock(&data->rx_lock);
+	ring_buf_get(&data->rx_rb, p_char, 1);
+	k_spin_unlock(&data->rx_lock, key);
+	return 0;
+}
+
+void uart_emul_poll_out(const struct device *dev, unsigned char out_char)
+{
+	struct uart_emul_data *data = dev->data;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&data->tx_lock);
+	ring_buf_put(&data->tx_rb, &out_char, 1);
+	k_spin_unlock(&data->tx_lock, key);
+}
+
+int uart_emul_err_check(const struct device *dev)
+{
+	return 0;
+}
+
+int uart_emul_configure(const struct device *dev, const struct uart_config *cfg)
+{
+	struct uart_emul_data *data = dev->data;
+
+	data->cfg.baudrate = cfg->baudrate;
+	data->cfg.data_bits = cfg->data_bits;
+	data->cfg.flow_ctrl = cfg->flow_ctrl;
+	data->cfg.parity = cfg->parity;
+	data->cfg.stop_bits = cfg->stop_bits;
+	return 0;
+}
+
+int uart_emul_config_get(const struct device *dev, struct uart_config *cfg)
+{
+	struct uart_emul_data *data = dev->data;
+
+	cfg->baudrate = data->cfg.baudrate;
+	cfg->data_bits = data->cfg.data_bits;
+	cfg->flow_ctrl = data->cfg.flow_ctrl;
+	cfg->parity = data->cfg.parity;
+	cfg->stop_bits = data->cfg.stop_bits;
+	return 0;
+}
+
+static const struct uart_driver_api uart_emul_api = {.poll_in = uart_emul_poll_in,
+						     .poll_out = uart_emul_poll_out,
+						     .config_get = uart_emul_config_get,
+						     .configure = uart_emul_configure,
+						     .err_check = uart_emul_err_check};
```

Comment:
Nit: use newline for struct initialisation
```suggestion
static const struct uart_driver_api uart_emul_api = {
	.poll_in = uart_emul_poll_in,
	.poll_out = uart_emul_poll_out,
	.config_get = uart_emul_config_get,
	.configure = uart_emul_configure,
	.err_check = uart_emul_err_check
};
```